src/main/java/org/folio/auth/authtokenmodule/impl/ModulePermissionsSource.java

```diff
@@ -73,70 +75,62 @@ public static void setCacheTimeout(int sec) {
-              logger.debug("Error parsing permissions object: " + e.getLocalizedMessage());
-              permissionsObject = null;
-            }
-            if (permissionsObject != null && permissionsObject.getJsonArray("permissionNames") != null) {
-              logger.debug("Got permissions");
-              promise.complete(permissionsObject.getJsonArray("permissionNames"));
-            } else {
-              logger.error("Got malformed/empty permissions object");
-              promise.fail("Got malformed/empty permissions object");
-            }
-          });
+          final String requestUrl = okapiUrl + "/perms/users/" + permUser.getString("id") + "/permissions?expanded=true";
+          logger.debug("Requesting permissions from URL at " + requestUrl);
+          HttpRequest<Buffer> req = client.getAbs(requestUrl);
+          endRequest(req, requestToken, tenant, requestId);
+          return req.send()
+              .compose(res -> {
+                if (res.statusCode() == 404) {
+                  //In the event of a 404, that means that the permissions user
+                  //doesn't exist, so we'll return an empty list to indicate no permissions
+                  return Future.succeededFuture(new JsonArray());
+                }
+                if (res.statusCode() != 200) {
+                  String failMessage = "Unable to retrieve permissions (code " + res.statusCode() + "): " + res.bodyAsString();
+                  logger.debug(failMessage);
+                  return Future.failedFuture(failMessage);
+                }
+                // 200
+                JsonObject permissionsObject;
+                try {
+                  permissionsObject = res.bodyAsJsonObject();
+                } catch (Exception e) {
+                  logger.debug("Error parsing permissions object: " + e.getLocalizedMessage());
+                  permissionsObject = null;
+                }
+                if (permissionsObject != null && permissionsObject.getJsonArray("permissionNames") != null) {
+                  logger.debug("Got permissions");
+                  return Future.succeededFuture(permissionsObject.getJsonArray("permissionNames"));
+                } else {
+                  logger.error("Got malformed/empty permissions object");
+                  return Future.failedFuture("Got malformed/empty permissions object");
+                }
+              });
         });
-        req.exceptionHandler(exception -> {
-          promise.fail(exception);
-        });
-        endRequest(req, requestToken, tenant, requestId);
-      });
-      permUserRes.exceptionHandler(e -> {
-        promise.fail(e);
-      });
-    });
-    permUserReq.exceptionHandler(promise::fail);
-    endRequest(permUserReq, requestToken, tenant, requestId);
-    return promise.future();
   }
 
-  private void endRequest(HttpClientRequest req, String requestToken,
+  private void endRequest(HttpRequest<Buffer> req, String requestToken,
```

Comment:
endRequest should be renamed, for example setHeaders

Reply:
Makes sense.